File: android/app/src/test/java/com/innocorelabs/verbale/data/UserItemsTest.kt
package com.innocorelabs.verbale.data

import com.innocorelabs.verbale.pipeline.Minutes
import com.innocorelabs.verbale.pipeline.Reconciler
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The two constants a hand-typed item rests on, pinned on the Kotlin side of the bridge.
 *
 * THESE ARE CROSS-LANGUAGE CONTRACTS, NOT TIDY CONSTANTS, which is what makes asserting a literal
 * worth doing. Both values are WRITTEN by JavaScript — `db.addUserItem` in src/db/queries.ts — and
 * READ here: `Reconciler` rule 1 refuses to match, replace or flag a row carrying
 * [AudioDb.Gen.USER], so a person's own item is never swallowed by an extracted one;
 * `AudioDb.indexItems` decides where a search hit on one opens the meeting;
 * `FileExportModule.exportItems` decides whether it is stamped with a moment.
 *
 * The two spellings meet only inside the database. Change one side and nothing fails to compile,
 * nothing throws, and no test anywhere else goes red: hand-typed items simply stop being
 * recognised as hand-typed, lose rule 1's protection, and are deleted by rule 4 on the first
 * reprocess — because the rules will never extract a sentence nobody said. That is the same class
 * of silent divergence `ItemKeyTest` exists for, and this is the same instrument: the literal,
 * asserted on both sides, in tests that name each other. The JavaScript half is
 * src/db/__tests__/userItems.test.ts.
 */
class UserItemsTest {

  /**
   * The gen a person's own item carries, spelled out.
   *
   * Unversioned on purpose, and the `else` branch in `db.allActions` is why: it reports anything
   * that is not exactly this string as `rule`, so a `user@1` — the convention `rules@1` sets for
   * the other half of this vocabulary — would reclassify every hand-typed action in the worklist
   * the day somebody added it.
   */
  @Test fun theUserGenIsSpelledTheWayJavaScriptWritesIt() {
    assertEquals("user", AudioDb.Gen.USER)
  }

  /** And it is not the rules' gen, which is the one thing it must never be mistaken for. */
  @Test fun theUserGenIsNotTheRulesGen() {
    assertNotEquals(Minutes.RULES_GEN, AudioDb.Gen.USER)
  }

  /**
   * The sentinel anchor is `Number.MAX_SAFE_INTEGER`, and that exact value is the point.
   *
   * Every query parameter crosses the bridge as JSON and `StorageModule.parseArgs` binds it as
   * text, so a sentinel above 2^53 arrives ROUNDED: `Long.MAX_VALUE` becomes 9223372036854775808,
   * which is not a valid INT64 and which SQLite therefore stores as a REAL in an INTEGER column —
   * after which the ordering this constant exists for is a floating-point comparison and the value
   * read back is not the value written. The round trip through a Double is what JSON does, so
   * asserting it survives one is the assertion, not `assertEquals` to a literal.
   */
  @Test fun theSentinelAnchorSurvivesTheJsonBridgeExactly() {
    assertEquals(9_007_199_254_740_991L, AudioDb.Gen.NO_ANCHOR)
    assertEquals(AudioDb.Gen.NO_ANCHOR, AudioDb.Gen.NO_ANCHOR.toDouble().toLong())
  }

  /**
   * And it sorts after anything that could ever be a real moment.
   *
   * Every read of `items` is `ORDER BY anchor_start_ms, rowid`, so this value is what keeps a
   * hand-typed row LAST — where it has always been, because it lived in `minutes` and was appended
   * after the items. A hundred years of continuous recording is the vector rather than a round
   * number, because the failure it guards against is a long recording, not an arithmetic one.
   */
  @Test fun theSentinelAnchorSortsAfterEveryRealMoment() {
    val aCenturyOfRecordingMs = 100L * 365 * 24 * 60 * 60 * 1000
    assertTrue(AudioDb.Gen.NO_ANCHOR > aCenturyOfRecordingMs)
  }

  /**
   * A row stored at the sentinel cannot be MATCHED, even with rule 1 out of the way.
   *
   * Rule 1 is what stops a hand-typed row reaching `Reconciler`'s matching at all, and it is meant
   * to be the only thing doing that work — so this asks what happens if it were ever deleted. The
   * row is labelled `rules@1` here precisely to get past rule 1 and reach rule 2's arithmetic:
   * `min(oldEnd, newEnd) - max(oldStart, newStart)`, kept only when it clears `MIN_OVERLAP_MS`.
   * With the sentinel that lands far below the threshold and, importantly, does not WRAP — a
   * subtraction that underflowed would come back hugely positive and the row would match
   * everything, which is the one failure a value this large could plausibly have.
   *
   * A PREVIOUS VERSION OF THIS TEST CLAIMED `0` "fails the other way", and that is false. It
   * asserted `overlap < 0`, which is the only reason `0` appeared to fail it. `MIN_OVERLAP_MS` is
   * 1, so a row stored at `0..0` yields `-newStart`, which is at most 0 and therefore also never
   * clears the threshold — the plan's own Task 7 note says so. The second half of this test is
   * that correction, asserted rather than written in a comment: both values are inert here, the
   * case for the sentinel over `0` is ORDERING and BRIDGE REPRESENTABILITY, and it does not need a
   * third leg that does not exist.
   *
   * Driven through `Reconciler.reconcile` rather than by re-deriving the subtraction, because the
   * threshold is `Reconciler`'s private business and a test that copies a private constant is a
   * test of its own copy. ReconcilerTest deliberately gives its user row a REAL anchor so that
   * rule 1 stays pinned as a `gen_version` rule rather than as an accident of this constant.
   */
  @Test fun neitherTheSentinelNorZeroCanEverBeMatched() {
    fun matchedIdOf(anchorStart: Long, anchorEnd: Long): String {
      val stored = AudioDb.StoredItem(
        "id-1", "action", "Send the report — Unassigned", AudioDb.Review.SUGGESTED,
        // NOT Gen.USER: rule 1 would set the row aside before the arithmetic below is reached, and
        // the arithmetic is the whole subject of this test.
        Minutes.RULES_GEN, anchorStart, anchorEnd, emptyList(), 1_772_000_000_000L, true,
      )
      val incoming = Minutes.Item(
        "action", "Send the report — Unassigned",
        listOf(Minutes.Source("u-new", 0L, 4_000L, 0, 5)), 0L, 4_000L,
      )
      val plan = Reconciler.reconcile(listOf(stored), listOf(incoming))
      // A match REUSES the stored id; a miss mints a fresh one and preserves the old row beside it.
      return plan.rows.first { it.item.anchorStartMs == 0L }.id
    }

    assertNotEquals(
      "the sentinel anchor wrapped and matched an item said at the start of the meeting",
      "id-1", matchedIdOf(AudioDb.Gen.NO_ANCHOR, AudioDb.Gen.NO_ANCHOR),
    )
    assertNotEquals(
      "0 was supposed to be the dangerous choice and is not — fix the claim, not the code",
      "id-1", matchedIdOf(0L, 0L),
    )
    // And the helper can tell a match from a miss. Without this the two assertions above pass
    // against a `matchedIdOf` that could never return "id-1" at all — a test of its own plumbing,
    // which is the single-row failure this branch keeps finding, wearing a different hat.
    assertEquals(
      "the same row at an anchor that really overlaps did not match, so the misses prove nothing",
      "id-1", matchedIdOf(0L, 4_000L),
    )
  }

  /**
   * The column is still NOT NULL, which is the entire reason a sentinel exists.
   *
   * If somebody ever rebuilds `items` to make `anchor_start_ms` nullable, the sentinel and every
   * derivation that turns it back into "no moment" become dead weight — and this is the test that
   * points at them, rather than leaving a value nobody can explain. SchemaTest's parser is the
   * precedent for reading the DDL as text on the JVM; the authoritative structural assertions live
   * in src/db/__tests__/schema.test.ts, which executes it.
   */
  @Test fun theAnchorColumnIsStillNotNullWhichIsWhyTheSentinelExists() {
    val items = AudioDb.schemaForTest().first { it.startsWith("CREATE TABLE IF NOT EXISTS items(") }
    assertTrue(
      "anchor_start_ms is nullable now — the sentinel and its derivations can go",
      items.contains("anchor_start_ms INTEGER NOT NULL"),
    )
    assertTrue(items.contains("anchor_end_ms INTEGER NOT NULL"))
  }
}

File: src/db/__tests__/userItems.test.ts
import { DatabaseSync } from 'node:sqlite';
import { SCHEMA } from '../schema';
import { NO_ANCHOR, USER_GEN } from '../../pipeline/types';

/**
 * Items a person typed themselves, written to and read back out of a real SQLite database.
 *
 * THE WHOLE TASK IS A MOVE BETWEEN TWO TABLES, so a test that mocked the result set would prove
 * only that the mock was written to match the assertion. These run the real statements from
 * `queries.ts` through `node:sqlite`, exactly as worklist.test.ts does, with the same caveat:
 * src/db/schema.ts is the READABLE copy of the DDL and `AudioDb.SCHEMA` is the one that runs.
 *
 * WHAT A HAND-TYPED ROW HAS TO SURVIVE, and why each assertion below is here rather than being
 * obvious. `items.anchor_start_ms` is `INTEGER NOT NULL` and SQLite cannot make a column nullable
 * without rebuilding the table, so a row that never claimed a moment still has to store a number.
 * Storing `0` would print `[0:00]` on an exported document and send a reader to the top of the
 * recording to look for a sentence nobody spoke — and, because every read of `items` is
 * `ORDER BY anchor_start_ms, rowid`, it would also silently move every hand-typed row to the TOP
 * of the MOM, which is not where it has ever been. So the column holds [NO_ANCHOR] and the fact
 * that means "none" is derived at this boundary from `gen_version`, never from the number.
 *
 * FIXTURES HERE HAVE TWO ROWS WHEREVER ORDER OR SELECTION IS ASSERTED, and they are arranged
 * adversarially: the hand-typed row is inserted FIRST and its text sorts FIRST alphabetically, so
 * a fixture cannot pass by accident under insertion order, under text order, or under the old
 * `anchor_start_ms = 0`.
 */
let mockSqlite: DatabaseSync;

jest.mock('../../native/NativeStorage', () => ({
  __esModule: true,
  default: {
    query: jest.fn(async (sql: string, params: string) => {
      const args = JSON.parse(params) as any[];
      const stmt = mockSqlite.prepare(sql);
      if (/^\s*SELECT/i.test(sql)) return JSON.stringify(stmt.all(...args));
      stmt.run(...args);
      return '[]';
    }),
    reindex: jest.fn(async () => {}),
  },
}));

// Imported after the mock so the module under test binds to the SQLite-backed stand-in.
import { db } from '../queries';
import Storage from '../../native/NativeStorage';

function ddlFor(name: string): string {
  const stmt = SCHEMA.find(
    s =>
      s.trimStart().startsWith(`CREATE TABLE IF NOT EXISTS ${name} (`) ||
      s.includes(`CREATE INDEX IF NOT EXISTS ${name} `),
  );
  if (!stmt) throw new Error(`no DDL for ${name} in SCHEMA`);
  return stmt;
}

function meeting(id: string, title: string, createdAt: number, archivedAt: number | null = null) {
  mockSqlite
    .prepare(
      'INSERT INTO meetings(id, title, created_at, duration_ms, status, archived_at) ' +
        "VALUES(?,?,?,0,'done',?)",
    )
    .run(id, title, createdAt, archivedAt);
}

/** An item the RULES produced, with a real moment behind it. */
function ruleItem(id: string, meetingId: string, kind: string, text: string, anchorStartMs: number) {
  mockSqlite
    .prepare(
      'INSERT INTO items(id, meeting_id, kind, text, review, gen_version, anchor_start_ms, ' +
        "anchor_end_ms, created_at) VALUES(?,?,?,?,'suggested','rules@1',?,?,0)",
    )
    .run(id, meetingId, kind, text, anchorStartMs, anchorStartMs + 1000);
}

function storedAnchors(id: string): { start: number; end: number } {
  const row = mockSqlite
    .prepare('SELECT anchor_start_ms AS start, anchor_end_ms AS end FROM items WHERE id = ?')
    .get(id) as unknown as { start: number; end: number };
  return row;
}

beforeEach(() => {
  mockSqlite = new DatabaseSync(':memory:');
  mockSqlite.exec('PRAGMA foreign_keys = ON;');
  for (const t of ['meetings', 'items', 'idx_items_meeting', 'item_sources', 'minutes']) {
    mockSqlite.exec(ddlFor(t));
  }
  (Storage.reindex as jest.Mock).mockClear();
});

afterEach(() => mockSqlite.close());

describe('addUserItem', () => {
  /**
   * A person wrote it, so there is nothing to review and nothing to cite.
   *
   * `review = 'confirmed'` is true by construction rather than a default anybody set, and
   * `gen_version = 'user'` is the ONE marker everything downstream keys on: Reconciler rule 1
   * refuses to match, replace or flag the row, `db.items` refuses it an anchor, and the export
   * refuses it a timestamp. Spelled from the shared constant on purpose — see USER_GEN.
   */
  it('is confirmed on arrival, cites nothing, and is marked as the person who wrote it', async () => {
    meeting('m1', 'Standup', 1000);

    const row = await db.addUserItem('m1', 'action', 'Book the venue — Priya');

    expect(row.review).toBe('confirmed');
    expect(row.sources).toEqual([]);
    expect(row.genVersion).toBe(USER_GEN);
    expect(row.kind).toBe('action');
    expect(row.text).toBe('Book the venue — Priya');
    expect(row.meetingId).toBe('m1');
    // The RETURNED row is what db.items would hand back, never what is on disk. Returning the
    // sentinel here survived every other assertion in this file: the return type is `Item`, whose
    // `anchorStartMs` is what a provenance button is gated on, so a caller that used it instead of
    // re-reading would draw a button offering to play a moment nobody spoke.
    expect(row.anchorStartMs).toBeNull();
    expect(row.anchorEndMs).toBeNull();
    // Searchable the moment it is typed, exactly as a hand-typed `minutes` row was.
    expect(Storage.reindex).toHaveBeenCalledWith('m1');
  });

  /**
   * The row it returns is the row it wrote. A method that returns a hand-built object the
   * database never saw is a method whose INSERT can be wrong in any way at all and still pass.
   */
  it('returns the row that is actually on disk', async () => {
    meeting('m1', 'Standup', 1000);
    ruleItem('i-rule', 'm1', 'action', 'Send the report — Unassigned', 65_000);

    const row = await db.addUserItem('m1', 'decision', 'Ship on the 14th');
    const back = (await db.items('m1')).find(i => i.id === row.id);

    expect(back).toBeDefined();
    expect(back!.kind).toBe('decision');
    expect(back!.text).toBe('Ship on the 14th');
    expect(back!.review).toBe('confirmed');
    expect(back!.genVersion).toBe(USER_GEN);
  });

  /**
   * Two typed in the same millisecond are two rows.
   *
   * `addUserMinute` already carries a random suffix for this reason; a bare `Date.now()` id makes
   * the second INSERT collide on the primary key, and a person adding two actions in one sitting
   * would see an error or lose one. Jest's clock is fast enough that this is not hypothetical.
   */
  it('mints a distinct id for two items typed in the same millisecond', async () => {
    meeting('m1', 'Standup', 1000);

    const a = await db.addUserItem('m1', 'action', 'Ring the supplier');
    const b = await db.addUserItem('m1', 'action', 'Book the venue');

    expect(a.id).not.toBe(b.id);
    expect((await db.items('m1')).map(i => i.text)).toEqual([
      'Ring the supplier',
      'Book the venue',
    ]);
  });
});

describe('the anchor of a row that never claimed a moment', () => {
  /**
   * The stored number is a sentinel, and the boundary turns it into `null`.
   *
   * Both halves matter and they fail differently. If the column held `0`, this assertion would
   * read `0` back and every renderer would print `[0:00]`. If the boundary forgot to map, the
   * screen would be handed 9007199254740991 and the provenance button — gated on
   * `anchorStartMs !== null` — would appear on a row with nothing to show.
   */
  it('stores a sentinel above every real anchor and hands back null', async () => {
    meeting('m1', 'Standup', 1000);
    ruleItem('i-rule', 'm1', 'action', 'Send the report — Unassigned', 65_000);

    const typed = await db.addUserItem('m1', 'action', 'Book the venue — Priya');

    expect(storedAnchors(typed.id)).toEqual({ start: NO_ANCHOR, end: NO_ANCHOR });
    const rows = await db.items('m1');
    expect(rows.find(i => i.id === typed.id)!.anchorStartMs).toBeNull();
    expect(rows.find(i => i.id === typed.id)!.anchorEndMs).toBeNull();
    // The extracted row keeps its own moment: a boundary that nulled EVERY anchor would pass an
    // assertion that only looked at the typed row.
    expect(rows.find(i => i.id === 'i-rule')!.anchorStartMs).toBe(65_000);
  });

  /**
   * The sentinel survives the bridge exactly.
   *
   * Every parameter crosses to Kotlin as JSON and comes back the same way, so a sentinel above
   * 2^53 would arrive rounded — `Long.MAX_VALUE` becomes 9223372036854775808, which SQLite stores
   * as a REAL in an INTEGER column. That is why the value is `Number.MAX_SAFE_INTEGER` rather
   * than the largest number the column could hold.
   */
  it('is exactly representable in JavaScript', async () => {
    meeting('m1', 'Standup', 1000);

    const typed = await db.addUserItem('m1', 'decision', 'Ship on the 14th');

    expect(NO_ANCHOR).toBe(Number.MAX_SAFE_INTEGER);
    expect(storedAnchors(typed.id).start).toBe(NO_ANCHOR);
    expect(Number.isSafeInteger(storedAnchors(typed.id).start)).toBe(true);
  });
});

describe('where a hand-typed row sits in the list', () => {
  /**
   * Last, which is where it has always been, and deliberately.
   *
   * Before this task a typed row lived in `minutes` and `toItemRows` appended it after every item,
   * so it came last no matter what. Moving it into `items` puts it under
   * `ORDER BY anchor_start_ms, rowid` for the first time, and `anchor_start_ms = 0` — the value
   * the plan's listing gives — would put it FIRST in every meeting, silently reversing Task 10's
   * deterministic order.
   *
   * The fixture is arranged so nothing else can produce this answer: the typed row is inserted
   * first, and "Book" sorts before "Send".
   */
  it('after every extracted row, however early it was typed', async () => {
    meeting('m1', 'Standup', 1000);

    const typed = await db.addUserItem('m1', 'action', 'Book the venue — Priya');
    ruleItem('i-late', 'm1', 'action', 'Send the report — Unassigned', 3_600_000);
    ruleItem('i-early', 'm1', 'decision', 'Ship on the 14th', 5_000);

    expect((await db.items('m1')).map(i => i.id)).toEqual(['i-early', 'i-late', typed.id]);
  });

  /** Two typed rows keep the order they were typed in, which is the only order they have. */
  it('and two typed rows keep the order they were typed in', async () => {
    meeting('m1', 'Standup', 1000);

    const first = await db.addUserItem('m1', 'action', 'Ring the supplier');
    const second = await db.addUserItem('m1', 'action', 'Book the venue');

    expect((await db.items('m1')).map(i => i.id)).toEqual([first.id, second.id]);
  });
});

describe('removeUserItem', () => {
  /**
   * Only ever the person's own row.
   *
   * The `gen_version` clause is the whole safety of this statement: an id arriving from a screen
   * that had drawn a stale list would otherwise delete an extracted item outright, and
   * `replaceItems` would not bring it back — the rules would, without its tick or its correction.
   */
  it('removes the row a person typed and refuses an extracted one', async () => {
    meeting('m1', 'Standup', 1000);
    ruleItem('i-rule', 'm1', 'action', 'Send the report — Unassigned', 65_000);
    const typed = await db.addUserItem('m1', 'action', 'Book the venue — Priya');

    await db.removeUserItem('m1', 'i-rule');
    expect((await db.items('m1')).map(i => i.id)).toEqual(['i-rule', typed.id]);

    await db.removeUserItem('m1', typed.id);
    expect((await db.items('m1')).map(i => i.id)).toEqual(['i-rule']);
  });

  /** The list it disappears from is also the index it disappears from. */
  it('reindexes the meeting', async () => {
    meeting('m1', 'Standup', 1000);
    const typed = await db.addUserItem('m1', 'action', 'Book the venue — Priya');
    (Storage.reindex as jest.Mock).mockClear();

    await db.removeUserItem('m1', typed.id);

    expect(Storage.reindex).toHaveBeenCalledWith('m1');
  });
});

describe('allActions', () => {
  /**
   * The cross-meeting worklist finally shows a hand-typed action — and shows it as one.
   *
   * Until this task an action somebody typed into a meeting was a `minutes` row with no item, so
   * it appeared on that meeting's own Actions tab and NOWHERE else: the one list whose job is to
   * say what you owe left it out. The anchor is null here for the same reason it is null in
   * `db.items`: `ActionRow.anchorStartMs` is what a future "take me to the moment" gesture reads,
   * and a hand-typed row has no moment to take anybody to.
   */
  it('includes a hand-typed action, marked as the person`s and with no moment', async () => {
    meeting('m1', 'Standup', 1000);
    ruleItem('i-rule', 'm1', 'action', 'Send the report — Unassigned', 65_000);
    const typed = await db.addUserItem('m1', 'action', 'Book the venue — Priya');

    const rows = await db.allActions();

    expect(rows.map(r => r.id)).toEqual(['i-rule', typed.id]);
    expect(rows[0]).toMatchObject({ source: 'rule', anchorStartMs: 65_000 });
    expect(rows[1]).toMatchObject({ source: 'user', anchorStartMs: null });
  });

  /** A typed DECISION is not an action, and the kind filter is what keeps it off the worklist. */
  it('leaves a hand-typed decision out', async () => {
    meeting('m1', 'Standup', 1000);
    ruleItem('i-rule', 'm1', 'action', 'Send the report — Unassigned', 65_000);
    await db.addUserItem('m1', 'decision', 'Ship on the 14th');

    expect((await db.allActions()).map(r => r.id)).toEqual(['i-rule']);
  });
});
